Multizone: initial docs

justinsb opened the pull request:
Initial pass on docs for multizone

justinsb commented:
TODO: Comment on original PR
> Also please be sure to explain whether the master components are running in a replicated/HA configuration, since I assume one of the primary reasons people might use Ubernetes Lite is to be able to survive zone failure.

justinsb commented:
cc @quinton-hoole

johndmulhausen commented on _data/guides.yml:
```diff
@@ -120,6 +120,8 @@ toc:
     path: /docs/admin/multi-cluster/
   - title: Using Large Clusters
     path: /docs/admin/cluster-large/
+  - title: Running in multiple zones
```
Caps for "Multiple Zones"

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+View the nodes; you can see that they are labeled with zone information.  They are all in `us-central1-a` (GCE) or `us-west-2a` (AWS) so far.  The labels are `failure-domain.beta.kubernetes.io/region` for the region, and `failure-domain.beta.kubernetes.io/zone` for the zone:
+
+```
````
84: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+Let's add another set of nodes to the existing cluster, reusing the existing master, running in a different zone (us-central1-b or us-west-2b).  We run kube-up again, but by specifying `KUBE_USE_EXISTING_MASTER=1` kube-up will not create a new master, but will reuse one that was previously created instead.
+
+GCE:
+```
````
100: (blank)
101: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+```
+
+On AWS we also need to specify the network CIDR for the additional subnet, along with the master internal IP address:
+```
````
105: (blank)
106: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+View the nodes again; 3 more nodes should have launched and be tagged in us-central1-b:
+
+```
````
112: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+Create a volume (only PersistentVolumes are supported for zone affinity), using the new dynamic volume creation:
+
+```
````
129: ```json

justinsb replied:
Technically it's shell, although it is primarily JSON.  Not sure whether there is a nice way to do this?

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+The PV is also labeled with the zone & region is was created in (which will currently always be the master zone, us-central1-a/us-west-2a):
+
+```
````
156: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+So now we will a pod that uses the persistent volume claim.  Because GCE PDs / AWS EBS volumes cannot be attached across volumes, this means that this pod can only be created in the same zone as the volume:
+
+```
````
164: ```yaml

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+And now we can verify that the pod was created in the same zone as the volume, as it had to be:
+
+```
````
186: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+
+GCE:
+```
````
200: (blank)
201: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+```
+
+AWS:
+```
````
205: (blank)
206: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+Verify that you now have nodes in 3 zones:
+
+```
````
211: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+Create the guestbook-go example, which includes an RC of size 3, running a simple web app:
+
+```
````
217: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+The pods should be spread across all 3 zones:
+
+```
````
223: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+Load-balancers can span zones; the guestbook-go example includes an example load-balanced service:
+
+```
````
239: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+```
+
+AWS:
+```
````
269: ```shell

johndmulhausen commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,273 @@
+
+GCE:
+
+```
````
262: ```shell

quinton-hoole commented:
@johndmulhausen I don't seem to have the required permissions to assign this to myself.  Can you fix that?

rata commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+pv-gce-mj4gm   5Gi        RWO           Bound     default/claim1             46s       failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-a
+```
+
+So now we will a pod that uses the persistent volume claim.  Because GCE PDs / AWS EBS volumes cannot be attached across volumes, this means that this pod can only be created in the same zone as the volume:
````
s/across volumes/across regions/?

quinton-hoole replied:
typo in first sentence.

quinton-hoole replied:
You might actually be able to omit this whole paragraph, and subsume it with my suggestion below.  Up to you.

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+## Introduction
+
+Kubernetes 1.2 adds support for running a single cluster in multiple zones.
```
"multiple availability zones" not "multiple zones"

quinton-hoole replied:
Possibly add something like "simply referred to as "zones" in the remainder of this document".

justinsb replied:
I think GCE refers to them as zones, so I was trying to not be AWS specific.  I wrote this:

> Kubernetes 1.2 adds support for running a single cluster in multiple failure zones
(GCE calls them simply "zones", AWS calls them "availability zones", here we'll refer to them as "zones").

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+Kubernetes 1.2 adds support for running a single cluster in multiple zones.
+This is a lightweight version of a broader effort for federating multiple
+Kubernetes clusters together (sometimes referred to by the affectionate
```
Good to add a link to further information here.
For now, I suggest that you use https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/federation.md
In a separate PR I will put together an index page linking to all of the Ubernetes docs, and update the link in this doc.

justinsb replied:
Adding link

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+Kubernetes clusters together (sometimes referred to by the affectionate
+nickname "Ubernetes").  Full federation will allow combining separate
+Kubernetes clusters running in different regions or clouds.  However, many
+users simply want to run a more available Kubernetes cluster in multiple zones
```
multiple availability zones

quinton-hoole replied:
Or perhaps alternatively follow suggestion above.

justinsb replied:
Going to use 'zones' throughout

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+Multizone support is deliberately limited: a single Kubernetes cluster can run
+in multiple zones, but only within the same region (and cloud provider).  Only
+GCE and AWS are currently supported automatically (though it would be easy to
```
Suggest rewording "would be" to "is".

justinsb replied:
Thanks

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+## Functionality
+
+When nodes are started, the kubelet will automatically add labels to them with
```
"will automatically add" => "automatically adds"

justinsb replied:
Fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+When nodes are started, the kubelet will automatically add labels to them with
+zone information.
+
+Just as pods in a replication controller or service are spread across nodes,
```
The wording of this sentence is confusing.  Suggest rewording to something like:

"Just as the pods in a replication controller or service are automatically spread across nodes in a single-zone cluster (to reduce the impact of node failures), this spreading behavior is extended across zones in a multi-zone cluster (to reduce the impact of zone failures).  This is achieved via `SelectorSpreadPriority`.

justinsb replied:
Sounds good.  Splitting the sentence into two for clarity.

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+Just as pods in a replication controller or service are spread across nodes,
+the scheduler will use the zone labels to spread first across nodes in
+different zones (via `SelectorSpreadPriority`).  This is a best-effort
+placement, and so if one zone has less available capacity you should not be
```
Actually less available capacity is fairly irrelevant here, I think, or at least slightly confusing.  It doesn't enter directly into the scheduling decision.  For example if I have A available capacity in one zone, and 2xA in another, I'm no more likely to land a pod in one or the other zone.  The only relevant thing is whether it is possible to schedule the required number of pods into a given zone to achieve equal spreading across zones.  I suggest rewording to something like:

".. and so if the zones in your cluster are heterogenous (e.g. different numbers of nodes, different types of nodes, or different pod resource requirements), this might prevent perfectly even spreading of your pods across zones.  If desired, you can use homogenous zones (same number and types of nodes) to reduce the probability of unequal spreading."

justinsb replied:
Using your text

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+placement, and so if one zone has less available capacity you should not be
+surprised if your pods are not scheduled evenly between the zones.
+
+When persistent volumes are created, labels are automatically added to them (by
```
"labels" => "zone labels"

justinsb replied:
Yes.  Also fixed passive voice.

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+When persistent volumes are created, labels are automatically added to them (by
+the `PersistentVolumeLabel` admission controller).  The scheduler (via the
+`VolumeZonePredicate` predicate) will then ensure that pods that claim that
+volume are only attached to volumes in their corresponding zone.
```
Some typos here: "will ensure that pods that claim a given volume are only placed into the same zone as that volume, as volumes cannot be attached across zones".

justinsb replied:
Fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+We assume that the different zones are located close to each other in the
+network, so we don't perform any zone-aware routing.  In particular, traffic
+that goes via services is likely to cross zones (even if pods in that service
```
"is likely" => "might"

justinsb replied:
Fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+We assume that the different zones are located close to each other in the
+network, so we don't perform any zone-aware routing.  In particular, traffic
+that goes via services is likely to cross zones (even if pods in that service
+are in the same zone), and this may incur additional latency and cost.
```
"even if some pods backing that service exist in the same zone as the client."

justinsb replied:
Fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+## Walkthough
+
+We're now going to walkthrough setting up and using a multi-zone cluster on both GCE & AWS.  The initial setup
```
walkthrough => walk through

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+## Walkthough
+
+We're now going to walkthrough setting up and using a multi-zone cluster on both GCE & AWS.  The initial setup
+is still a little awkward, in that you bring up a full cluster (specifying `MULTIZONE=1`),
```
I would omit the apology.  I don't think that it's particularly awkward, and in fact is extremely useful if one wants to add zones to a cluster, which will, I think, be fairly common.

justinsb replied:
OK!  I hope we can still make this better with some of mike danese's work, but removing the pre-emptory apology!

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+We're now going to walkthrough setting up and using a multi-zone cluster on both GCE & AWS.  The initial setup
+is still a little awkward, in that you bring up a full cluster (specifying `MULTIZONE=1`),
+and then you add nodes in additional zones by running `kube-up` again (specifying `KUBE_USE_EXISTING_MASTER=true`).
+Hopefully this will be improved in 1.3, and in any case once it is running the experience is much nicer.
```
Omit this sentence.

justinsb replied:
Fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+GCE:
+
+```shell
+curl -sS https://get.k8s.io | MULTIZONE=1 KUBE_GCE_ZONE=us-central1-a bash
````
Perhaps add KUBERNETES_PROVIDER=gce here (even though it's the default), to make clear the symmetry between the GCE and AWS commands.

justinsb replied:
Ooh, that is nice.  Fixing.

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+GCE:
+
+```shell
+KUBE_USE_EXISTING_MASTER=true MULTIZONE=1 KUBE_GCE_ZONE=us-central1-b kubernetes/cluster/kube-up.sh
````
nit: I don't quite remember how the script knows where to find the existing master?  It might be worth mentioning that briefly here, to avoid confusion.  Up to you.

justinsb replied:
Well it's actually by name or well-known IP.  I think it might be better to gloss over this detail, as the script doesn't really find the existing master...

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+```
+
+
+View the nodes again; 3 more nodes should have launched and be tagged in us-central1-b:
````
nit: It might be worth adding the NODE_COUNT (or whatever) parameter to the above commands, otherwise "3" seems like a magic number here?

justinsb replied:
Good point.  Not least because the default is 4 and I just ran with 3 to keep it sensible!

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+EOF
+```
+
+The PV is also labeled with the zone & region is was created in (which will currently always be the master zone, us-central1-a/us-west-2a):
````
typo: "is"->"it".
We should add this restriction to the Limitations section.  "It is currently only possible to create persistent volumes in the  zone of the cluster master" or some such, ideally with some indication that we plan to fix this soon.

quinton-hoole replied:
Is there an issue to track fixing this?  It's a pretty horrible limitation.  To avoid changing the API just yet, we could perhaps use a label or annotation on the PV claim?  Much like https://github.com/kubernetes/kubernetes/pull/22985/files/6c9c434f8aa5019d966717aad33955e6c5dcc900#r56701916.  cc @saad-ali for advice, in case he has any.

justinsb replied:
It's only _dynamic_ PVs - if you create the manually you can do whatever you want.  But I agree it is not ideal.  I have opened https://github.com/kubernetes/kubernetes/issues/23330 and linked to it in the docs.

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+EOF
+```
+
+And now we can verify that the pod was created in the same zone as the volume, as it had to be:
````
nit: Perhaps reword to something like "Note that the pod was created in the same zone as the volume, as cross-zone attachments are not generally permitted by cloud providers".

justinsb replied:
Using your text

quinton-hoole commented on docs/admin/multiple-zones.md:
```diff
@@ -0,0 +1,280 @@
+
+### Pods are spread across zones
+
+Pods in a replication controller or service are automatically spread across zones.  First, let's launch more pods in a third zone:
```
"launch more nodes"

justinsb replied:
Thanks.  Please send coffee to prevent further similar mistakes ;-)

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+KUBE_USE_EXISTING_MASTER=true MULTIZONE=1 KUBERNETES_PROVIDER=aws KUBE_AWS_ZONE=us-west-2c KUBE_SUBNET_CIDR=172.20.2.0/24 MASTER_INTERNAL_IP=172.20.0.9 kubernetes/cluster/kube-up.sh
+```
+
+Verify that you now have nodes in 3 zones:
````
Same comment as above regarding "3".

justinsb replied:
fixing

quinton-hoole commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,280 @@
+```
+
+
+Load-balancers can span zones; the guestbook-go example includes an example load-balanced service:
````
nit: omit "can".  Perhaps just "Load-balancers span all zones in a cluster".

justinsb replied:
fixing

quinton-hoole commented:
@johndmulhausen I don't have permission to add ("LGTM") or remove ("Needs Tech Review") labels in this new docs repo.  @bgrant0607 is that by design or accident?  Please advise on what process you'd like us to follow to get this doc merged.  Thanks.

@justinsb:
This is great! The structure and coverage are perfect, IMO. Many thanks!
Comments are of a relatively minor nature.  Once addressed to your satisfaction, I consider this tech review complete, and we can merge.

bgrant0607 commented:
@quinton-hoole At the moment, by design. @johndmulhausen will review docs for formatting, etc.

Hopefully we'll get more automation in place soon to help with this.

quinton-hoole commented:
@justinsb one more thing.  Still need to add reference to the fact that kube-up still builds non-HA masters by default, and that users should follow "high-availability.md", spreading across zones if they want their API to be fully tolerant to zone outages.  But with or without that, services are tolerant of zone failures.

quinton-hoole commented:
@bgrant0607 @johndmulhausen could one of you assign this to someone to get it merged ASAP.  Just want to make sure that it doesn't fall off the radar.  Thanks.

johndmulhausen commented:
@quinton-hoole It sounds like there are still tech review issues, are there not? Just let me know the tech review is done by labeling this "Tech Review: LGTM" and I'll merge it.

justinsb commented:
Added fixes per review.

We probably should merge this if there's nothing egregious given we've released 1.2 (delay was my fault though!), and then we can fix remaining issues with more PRs.

@johndmulhausen how do we feel about shell commands that stretch across the full width of a widescreen monitor?  Is it better to break them with `\` and risk people copying and pasting incorrectly?

justinsb commented:
PS thanks for the detailed review @quinton-hoole

quinton-hoole commented:
@johndmulhausen please merge this ASAP. As I've mentioned before, I do not have permission to change labels on any issues in this repo, so I cannot add the "Tech Review: LGTM" label per your request. You will have to do that for me until the permissions are changed. Thanks.
cc @bgrant0607 FYI

bgrant0607 commented on docs/admin/multiple-zones.md:
````diff
@@ -0,0 +1,313 @@
+```shell
+> kubectl get nodes --show-labels
+
+NAME                     STATUS                     AGE       LABELS
````
@johndmulhausen Is shell really the recommended highlighter for command output, or only for the command?

And are we really using `>` for the prompt convention? That's totally different than ~all other docs AFAIK.

justinsb replied:
What should I use?  $ ?

bgrant0607 replied:
@pwittrock What prompt did you settle on for hellonode? None?

pwittrock replied:
As written, hellonode doesn't use a prompt and drops the output right after the command.  I have been doing a bit of searching on solutions, but am not certain what the right thing to do here is.

As one data point, [this style guild](http://www.cirosantilli.com/markdown-style-guide/#dollar-signs-in-shell-code) recommends no-prompt if command output is not shown, and using a prompt if command output is shown.

pwittrock replied:
Alternatively, we could use block quotes for command output instead of code blocks:

```shell command```
>
<i>command
output</i>

justinsb replied:
@pwittrock I tried this suggestion (I like it), but I couldn't make the quoted output appear in a fixed-width font, so it looked weird because the tables weren't aligned.  Is there a trick I'm missing?

I was thinking that it would be cool to have something like you just described, with
```
<div>
<pre style="command">open the pod bay doors</div>
<pre style="response">
I'm sorry Dave.
I can't do that
</pre>
</div>
```

where command would be selectable (and thus easily copied), but the output would not be, and likely in a slight different color.  But I suspect we're heading pretty far out of the range of what github pages can achieve at this point...

pwittrock replied:
@justinsb I haven't experimented much with the command formatting.  What you suggest sounds awesome, but beyond my html expertise.  @johndmulhausen may be able to give better guidance in this area.

bgrant0607 commented:
Please take a look at:
https://github.com/kubernetes/kubernetes.github.io/blob/master/README.md

It would be useful to rename the fork so that we could view the content on justinsb.github.io.

justinsb commented:
I renamed the fork and removed the CNAME, so it now shows.  But...

1. It is showing master - any way to switch?
1. I suspect you're going to get a lot of PRs which accidentally delete CNAME
1. I expect a lot of people aren't going to want to dedicate their github.io page to this
1. Do we have canonical links set so that we don't get penalized for duplicate pages?

pwittrock commented:
@justinsb FWIW I remove the CNAME from master branch and merge changes from my dev branch into the master branch.  It isn't ideal, but helps address 1 & 2.  It does make it hard to have multiple PRs out for review at the same time.

justinsb commented:
Squashed.  Also pushed to http://justinsb.github.io/docs/admin/multiple-zones/ per your trick @pwittrock  (thanks).  Although we might do better to have a docs PR builder, as I think this process is very likely to have problems!

justinsb commented:
There's probably about a user a day asking on slack about how to do this, which suggests there are more that don't even get that far.  Can we just merge this as-is, and then fix the formatting of the output later?  (I don't think there are other outstanding issues?)

pwittrock commented:
I am good with fixing the formatting of the output later.

bgrant0607 commented:
Ok, merge and iterate.
